network/hits.go

```diff
@@ -0,0 +1,101 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package network
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/graph"
+)
+
+// HubAuthority is a Hyperlink-Induced Topic Search hub-authority score pair.
+type HubAuthority struct {
+	Hub       float64
+	Authority float64
+}
+
+// HITS returns the Hyperlink-Induced Topic Search hub-authority scores for
+// nodes of the directed graph g. HITS terminates when the 2-norm of the
+// vector difference between iterations is below eps. The returned map is
+// keyed on the graph node IDs.
+func HITS(g graph.DirectedGraph, eps float64) map[int]HubAuthority {
+	nodes := g.NodeList()
+
+	// Make a topological copy of g with dense node IDs.
+	indexOf := make(map[int]int, len(nodes))
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+	to := make([][]int, len(nodes))
```

Comment:
Could we call these "into" and "outfrom" or something like that? The current terminology is confusing because "to" are the nodes that point into i, instead of the nodes that i points to.